libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -116,17 +117,23 @@ void ContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinition const&
 			);
 	}
 
-	for (auto const* ancestorContract: _contract.annotation().linearizedBaseContracts | ranges::views::reverse)
+	auto const& contractHierarchy = _contract.annotation().linearizedBaseContracts;
+	for (auto const& baseContractSpecifier: _contract.baseContracts())
 	{
-		if (*ancestorContract == _contract)
-			continue;
-		if (ancestorContract->storageLayoutSpecifier())
+		std::string baseName = baseContractSpecifier->name().path().back();
+		auto baseContract = ranges::find_if(
+			contractHierarchy,
+			[&](auto const& _ancestorContract) { return baseName == _ancestorContract->name(); }
+		);
+		solAssert(baseContract != ranges::end(contractHierarchy));
+
+		if ((*baseContract)->storageLayoutSpecifier())
 			m_errorReporter.typeError(
 				8894_error,
-				_contract.location(),
+				baseContractSpecifier->location(),
 				SecondarySourceLocation().append(
 					"Storage layout was already specified here.",
-					ancestorContract->storageLayoutSpecifier()->location()
+					(*baseContract)->storageLayoutSpecifier()->location()
 				),
 				"Storage layout can only be specified in the most derived contract."
```

Comment:
I'd adjust the error messages now that we point at the layout specifier rather than the most derived contract:

```suggestion
					"Custom storage layout defined here:",
					(*baseContract)->storageLayoutSpecifier()->location()
				),
				"Cannot inherit from a contact with a custom storage layout."
```

Reply:
Done.